spark/spark-2.0/src/test/scala/org/apache/spark/sql/hive/HivemallOpsSuite.scala

```diff
@@ -649,6 +686,66 @@ final class HivemallOpsWithFeatureSuite extends HivemallFeatureQueryTest {
     val row4 = df4.groupby($"c0").f1score("c1", "c2").collect
     assert(row4(0).getDouble(1) ~== 0.25)
   }
+
+  test("user-defined aggregators for ftvec.selection") {
+    import hiveContext.implicits._
+    implicit val doubleEquality = org.scalactic.TolerantNumerics.tolerantDoubleEquality(1e-5)
+
+    // see also hivemall.ftvec.selection.SignalNoiseRatioUDAFTest
+    // binary class
+    // +-----------------+-------+
+    // |     features    | class |
+    // +-----------------+-------+
+    // | 5.1,3.5,1.4,0.2 |     0 |
+    // | 4.9,3.0,1.4,0.2 |     0 |
+    // | 4.7,3.2,1.3,0.2 |     0 |
+    // | 7.0,3.2,4.7,1.4 |     1 |
+    // | 6.4,3.2,4.5,1.5 |     1 |
+    // | 6.9,3.1,4.9,1.5 |     1 |
+    // +-----------------+-------+
+    val df0 = Seq(
+      (1, Seq(5.1, 3.5, 1.4, 0.2), Seq(1, 0)), (1, Seq(4.9, 3.0, 1.4, 0.2), Seq(1, 0)),
+      (1, Seq(4.7, 3.2, 1.3, 0.2), Seq(1, 0)), (1, Seq(7.0, 3.2, 4.7, 1.4), Seq(0, 1)),
+      (1, Seq(6.4, 3.2, 4.5, 1.5), Seq(0, 1)), (1, Seq(6.9, 3.1, 4.9, 1.5), Seq(0, 1)))
+      .toDF("c0", "arg0", "arg1")
+    val row0 = df0.groupby($"c0").snr("arg0", "arg1").collect
+    (row0(0).getAs[Seq[Double]](1), Seq(4.38425236, 0.26390002, 15.83984511, 26.87005769))
+      .zipped
+      .foreach((actual, expected) => actual shouldEqual expected)
+
+    // multiple class
+    // +-----------------+-------+
+    // |     features    | class |
+    // +-----------------+-------+
+    // | 5.1,3.5,1.4,0.2 |     0 |
+    // | 4.9,3.0,1.4,0.2 |     0 |
+    // | 7.0,3.2,4.7,1.4 |     1 |
+    // | 6.4,3.2,4.5,1.5 |     1 |
+    // | 6.3,3.3,6.0,2.5 |     2 |
+    // | 5.8,2.7,5.1,1.9 |     2 |
+    // +-----------------+-------+
+    val df1 = Seq(
+      (1, Seq(5.1, 3.5, 1.4, 0.2), Seq(1, 0, 0)), (1, Seq(4.9, 3.0, 1.4, 0.2), Seq(1, 0, 0)),
+      (1, Seq(7.0, 3.2, 4.7, 1.4), Seq(0, 1, 0)), (1, Seq(6.4, 3.2, 4.5, 1.5), Seq(0, 1, 0)),
+      (1, Seq(6.3, 3.3, 6.0, 2.5), Seq(0, 0, 1)), (1, Seq(5.8, 2.7, 5.1, 1.9), Seq(0, 0, 1)))
+      .toDF("c0", "arg0", "arg1")
+    val row1 = df1.groupby($"c0").snr("arg0", "arg1").collect
+    (row1(0).getAs[Seq[Double]](1), Seq(8.43181818, 1.32121212, 42.94949495, 33.80952381))
+      .zipped
+      .foreach((actual, expected) => actual shouldEqual expected)
+  }
+
+  test("user-defined aggregators for tools.matrix") {
+    import hiveContext.implicits._
+
+    // | 1  2  3 |T    | 5  6  7 |
+    // | 3  4  5 |  *  | 7  8  9 |
+    val df0 = Seq((1, Seq(1, 2, 3), Seq(5, 6, 7)), (1, Seq(3, 4, 5), Seq(7, 8, 9)))
+      .toDF("c0", "arg0", "arg1")
+
+    df0.groupby($"c0").transpose_and_dot("arg0", "arg1").collect() shouldEqual
+      Seq(Row(1, Seq(Seq(26.0, 30.0, 34.0), Seq(38.0, 44.0, 50.0), Seq(50.0, 58.0, 66.0))))
+  }
```

Comment:
@maropu oh, really? If use the same test on spark-1.6, it fail with following error

```
- user-defined aggregators for tools.matrix *** FAILED ***
  AttributeSet(hiveudaffunction(HiveFunctionWrapper(hivemall.tools.matrix.TransposeAndDotUDAF,hivemall.tools.matrix.TransposeAndDotUDAF@526ee5e6),arg0#456,arg1#457,false,0,0)#460) was not empty The physical plan has missing inputs: SortBasedAggregate(key=[c0#455], functions=[(hiveudaffunction(HiveFunctionWrapper(hivemall.tools.matrix.TransposeAndDotUDAF,hivemall.tools.matrix.TransposeAndDotUDAF@526ee5e6),arg0#456,arg1#457,false,0,0),mode=Complete,isDistinct=false)], output=[c0#455,hiveudaffunction(HiveFunctionWrapper(hivemall.tools.matrix.TransposeAndDotUDAF,hivemall.tools.matrix.TransposeAndDotUDAF@526ee5e6),arg0,arg1,false,0,0)#459])
  +- ConvertToSafe
     +- Sort [c0#455 ASC], false, 0
        +- TungstenExchange hashpartitioning(c0#455,5), None
           +- ConvertToUnsafe
              +- LocalTableScan [c0#455,arg0#456,arg1#457], [[1,[1,2,3],[5,6,7]],[1,[3,4,5],[7,8,9]]] (QueryTest.scala:197)
R
```